cloud/ec2.go

```diff
@@ -1167,66 +1167,57 @@ func (m *ec2Manager) TerminateInstance(ctx context.Context, h *host.Host, user,
 	if h.Status == evergreen.HostStopped {
 		return errors.Errorf("cannot stop '%s' - already marked as stopped", h.Id)
-	} else if h.Status != evergreen.HostRunning {
-		return errors.Errorf("cannot stop '%s' - host is not running", h.Id)
+	} else if h.Status != evergreen.HostRunning && h.Status != evergreen.HostStopping {
+		return errors.Errorf("cannot stop '%s' - host status is '%s'", h.Id, h.Status)
 	}
 
 	if err := m.client.Create(m.credentials, m.region); err != nil {
-		return errors.Wrap(err, "error creating client")
+		return errors.Wrap(err, "creating client")
 	}
 	defer m.client.Close()
 
-	var instance *ec2.Instance
-	instance, err := m.client.GetInstanceInfo(ctx, h.Id)
-	if err != nil {
-		return errors.Wrap(err, "error getting instance info before stopping")
-	}
-
-	// if instance is already stopped, return early
-	if ec2StatusToEvergreenStatus(*instance.State.Name) == StatusStopped {
-		grip.Info(message.Fields{
-			"message":       "instance already stopped",
-			"user":          user,
-			"host_provider": h.Distro.Provider,
-			"host_id":       h.Id,
-			"distro":        h.Distro.Id,
-		})
-
-		return errors.Wrap(h.SetStopped(user), "failed to mark instance as stopped in db")
-	}
-
-	if err = h.SetStopping(user); err != nil {
-		return errors.Wrap(err, "failed to mark instance as stopping in db")
-	}
-
-	_, err = m.client.StopInstances(ctx, &ec2.StopInstancesInput{
+	out, err := m.client.StopInstances(ctx, &ec2.StopInstancesInput{
 		InstanceIds: []*string{aws.String(h.Id)},
 	})
 	if err != nil {
-		// fetch the instance to get the current status, since it may have changed
-		// since we saved the previous status
-		currentStatus, currentStatusErr := m.GetInstanceStatus(ctx, h)
-		if currentStatusErr != nil {
-			return errors.Wrap(currentStatusErr, "error getting instance status after stopping error")
-		}
-
-		if currentStatusErr = h.SetStatusAtomically(currentStatus.String(), user, "reverting to the most up-to-date host status"); currentStatusErr != nil {
-			return errors.Wrapf(currentStatusErr, "failed to revert status from stopping to '%s'", currentStatus)
-		}
-		return errors.Wrapf(err, "error stopping EC2 instance '%s'", h.Id)
+		return errors.Wrapf(err, "stopping EC2 instance '%s'", h.Id)
 	}
 
-	// Check whether instance stopped
+	if len(out.StoppingInstances) == 1 {
+		instance := out.StoppingInstances[0]
+		status := ec2StatusToEvergreenStatus(utility.FromStringPtr(instance.CurrentState.Name))
+		switch status {
+		case StatusStopping:
+			grip.Error(message.WrapError(h.SetStopping(user), message.Fields{
+				"message": "could not mark host as stopping, continuing to poll instance status anyways",
+				"host_id": h.Id,
+				"user":    user,
+			}))
+		case StatusStopped:
+			return errors.Wrap(h.SetStopped(user), "marking DB host as stopped")
+		default:
+			return errors.Errorf("instance is in unexpected state '%s'", status)
+		}
+	}
+	grip.WarningWhen(len(out.StoppingInstances) != 1, message.Fields{
```

Comment:
if this is the case is it worth going through with the rest of the function to poll the instance status waiting for it to stop?

Reply:
This shouldn't happen unless we write a bug in the code, AWS returns an unexpected result, or our understanding of the AWS API results is incorrect. This is more of a sanity check since AWS is allowed to return us a slice of any number of results. The API should return exactly 1 result if it succeeded, but we can't guarantee that on our end. Checking `out.StoppingInstances` is just a shortcut to avoid an unnecessary polling loop if AWS tells us it's already stopped.

In the case of unexpected number of results, polling should still probably be done (at least from a correctness perspective) so that `GetInstanceInfo` can tell us what the real cloud host status is after we've received an unexpected number of results to make it more likely that the cloud host and DB host statuses are in sync with each other.